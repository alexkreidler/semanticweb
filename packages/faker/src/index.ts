import { NamedNode, Dataset } from "rdf-js";

/* # Input
* This section defines the input structure of the API

* BasicInput requires all the fields that are needed to process a requested fake document
* All fields to generate should simply contain true, but the type and context may optionally be strings. The input will be parsed with JSON-LD to get the fully normalized requested value
*/
export type BasicInput = { [key: string]: true } & (
  | { "@type": string }
  | { "@type": string; "@context": string }
);

// Mapping
// This sectinon of the API allows the core library to map simple values such as schema:email to generator functions

type FunctionMap = { [key: string]: Function };

// The simplified map is how all of our data is represented at first. Then it is processed by adding the prefix to the key of the function map.
type SimplifiedMap = { [prefix: string]: FunctionMap };

export interface MapProcessor {
  // This function adds the prefix to the key of the function map.
  expand(input: SimplifiedMap): TypeGeneratorMap;
}

// TypeGeneratorMap is a map from fully-qualified Semantic Web IRIs to a function which generates representative data
export type TypeGeneratorMap = { [semanticType: string]: Function };

// Complex properties
// These will only be tried if the simply properties have failed
// For thesse, the system needs a way to determine the effective "type" of the property, which it can then pass to the simple MapProcessor
// It needs support for dereferencing these IRIs

// Implementation note: for now, we don't want to hard code in rdf-vocabularies if possible. Let's just encode a few core ones like
// http://schema.org/rangeIncludes and similar

export interface ComplexPropertyProcessor {
  node: NamedNode;

  // fetches the data from the given IRI
  fetchData(): Promise<Dataset>;
  readonly data: Dataset;

  // returns the effective type of the node
  // this can then be looked up if in the mapping if it is a simple value
  // if not, it needs to be represented
  getType(): NamedNode;
}

export const ex: TypeGeneratorMap = {};
export const inp: FunctionMap = {}; 
// interface

// export interface Generator {
//   semanticMap;
// }
