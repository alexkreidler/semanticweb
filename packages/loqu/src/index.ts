import { Frame, JsonLd } from "./jsonld-types"

import * as jsonld from "jsonld"
import { Options } from "jsonld"

enum Strictness {
    /** This just frames the data but doesn't impose any additional options */
    NoChecks,
    /** Frame strict mode enables both the Explicit Inclusion and the Require All
     * flags on the Frame. How to deal with no-data/error: a validation strategy? */
    FrameStrict,

    /** This checks that XSD data types are properly formatted */
    BasicValidation,

    /** This allows for custom validation using the shacl and other options. Not implemented */
    AdvancedValidation,
}
enum Conversion {
    None,
    /** Basic conversion converts xsd data types to JS */
    BasicConversion,
}

/**
 * API Module boundary
 */
export type SFunc<O, I = JsonLd> = (data: I) => O
/** Do we want to make this generic over return types? */
export type SemanticFunction<T> = {
    data: {
        strictness: Strictness
        conversion: Conversion
        frame: {
            spec: Frame
            opts: Options.Frame
        }
    }
    /** In this case, the function doesn't have a return value as this
     * isn't a generation, validation, or conversion processing node */
    func: SFunc<T>
}

export function defaultSemanticFunction(): SemanticFunction<void> {
    return {
        data: {
            // what should this default be
            strictness: Strictness.FrameStrict,
            conversion: Conversion.None,
            frame: {
                spec: {},
                opts: {},
            },
        },
        func: (data: JsonLd) => {},
    }
}

export function createSemanticFunction<T>(frame: Frame, func: SFunc<T>): SemanticFunction<T> {
    let sf = defaultSemanticFunction()
    sf.data.frame.spec = frame
    sf.func = func
    // This assertion only works because we set func to func. This needs to stay coupled with
    // this set of APIs
    return sf as SemanticFunction<T>
}
/**
 * End API Module boundary
 */

export async function runSingleFunc<T>(sf: SemanticFunction<T>, data: JsonLd): Promise<T> {
    /* Prepare frame options */
    let frameOpts: Options.Frame = {}
    switch (sf.data.strictness) {
        case Strictness.NoChecks:
            frameOpts = {}
            break

        case Strictness.FrameStrict:
            frameOpts = {
                requireAll: true,
                explicit: true,
            }
            break

        default:
            console.log("Currently other validation types have not been implemented")

            break
    }
    /* We merge in the explicitly provided options, setting them to override */
    const realFrameOpts = { ...frameOpts, ...sf.data.frame.opts }

    const framed = await jsonld.frame(data, sf.data.frame.spec, realFrameOpts)

    const output = sf.func(framed)
    return output
}

export * from "./genericHelpers"
export * from "./rdfineHelpers"

export type IDNode = {
    "@id": string
}
