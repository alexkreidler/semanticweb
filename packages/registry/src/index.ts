export { IComponentRegistry, ComponentRegistry } from "./registry"

export { Component, Registry } from "./components"
