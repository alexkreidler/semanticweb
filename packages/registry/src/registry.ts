import React from "react"

import { ok, err, ResultAsync, Result } from "neverthrow"
// TODO: reevaluate if this poses too much API burden on consumers

type Empty = {}
type IRI = string

type ParseError = {
    message: "Invalid IRI"
}

type InvalidComponentError = {
    message: "Invalid Component"
}

type NotFoundError = {
    message: "Component Not Found"
}

type Error = ParseError | NotFoundError | InvalidComponentError

export interface IComponentRegistry<DataProps> {
    registerCallback: () => void
    handleCallback: () => void
    register(iri: IRI, component: React.ReactNode): Result<Empty, Error>
    // this is called from the context of the @type field in a data object
    // however, the system should expand to use @base and @vocab

    // undefined is a valid react node. Should we fail silently??
    handle(iri: IRI): Result<React.FC<DataProps>, NotFoundError>
    // Result<React.ReactNode, NotFoundError>
}

const normalizeIRI = (iri: string): string => {
    console.debug("input iri", iri)
    // TODO: think about more complex normalization
    let out = iri.replace(/^http:\/\//i, "https://")
    console.debug("output iri", out)
    return out
}

export class ComponentRegistry<DataProps>
    implements IComponentRegistry<DataProps> {
    private map: { [key: string]: React.FC<DataProps> } = {}
    constructor() {}
    registerCallback: () => void = () => {}
    handleCallback: () => void = () => {}
    register(
        iri: string,
        component: React.FC<DataProps>
    ): Result<Empty, Error> {
        // TODO: validate IRI
        iri = normalizeIRI(iri)

        if (component == undefined) {
            return err({ message: "Invalid Component" })
        }

        this.map[iri] = component

        this.registerCallback()
        return ok({})
    }
    handle(iri: string): Result<React.FC<DataProps>, NotFoundError> {
        iri = normalizeIRI(iri)
        // TODO: should require full IRI?
        let component = this.map[iri]
        if (component == undefined) {
            return err({ message: "Component Not Found" })
        }
        this.handleCallback()
        return ok(component)
    }
}
