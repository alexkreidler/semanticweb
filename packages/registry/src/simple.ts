export type SimpleRegistry = {
    /// The key should be the normalized IRI
    [key: string]: RegistryItem
}

/// A registry item is either a function (for JIT module loading) or a react node to render
export type RegistryItem = Function | React.ReactNode

// TODO: maybe add more restrictions on the proptypes of this node, to explicitly include sweb data

// think about cyclic data
