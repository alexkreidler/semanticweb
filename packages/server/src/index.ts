// TODO: dynamically load and bootstrap all configured frontends  and backends

// import { BasicServer } from "./server/server"

// BasicServer.start()
