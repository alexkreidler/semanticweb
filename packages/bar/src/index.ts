import { version } from '@semanticweb/faker';

console.log(version);
