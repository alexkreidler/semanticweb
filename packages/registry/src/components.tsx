import React, { useState, useEffect, useContext, createContext } from "react"

import { ComponentRegistry } from "./registry"

import jsonld from "jsonld"

// type Data = {
//     "@type": string
//     [key: string]: any
// }

// TODO: make data type passed in then map to the @type string
// maybe if we require expanded data to be passed this would work
export type RegistryProps = {
    /** The input data. Should be in (expanded) JSON-LD format */
    data: any
    children: React.ReactElement<ComponentProps>[]

    /** For now, we just require it to be expanded beforehand. This doesn't do anything */
    doExpand?: boolean
}

let ctx = createContext({ __SEMANTIC_REGISTRY_EMPTY: true })

export const Registry: React.FC<RegistryProps> = ({ children, data }) => {
    console.log({ children })

    let reg = new ComponentRegistry<MinimalChildProps>()
    let c = useContext(ctx)

    if (Object.keys(data).length !== 1) {
        return (
            <>
                Error in Semantic Web Registry: Data is not properly formatted
                in expanded format
            </>
        )
    }

    let obj = data[Object.keys(data)[0]]
    // we only take the first defined type.
    // TODO: fix??
    let fullTypeIRI = obj["@type"][0]

    React.Children.map(children, (c) => {
        if (!c.props.children && !c.props.component) {
            throw new Error(
                "Provided <Component> doesn't specify a child component to render"
            )
        } else if (c.props.children && c.props.component) {
            throw new Error(
                "Provided <Component> specifies two child components"
            )
        } else {
            let child = c.props.component ? c.props.component : c.props.children
            reg.register(c.props.iri, child)
        }
    })

    // let res
    let res = reg.handle(fullTypeIRI)

    if (res.isErr()) {
        return <>Error in Semantic Web Registry: {res.error.message}</>
        // return <>Error in Semantic Web Registry: {res.error.message}</>
    } else {
        let Final = res.value
        return (
            <>
                <Final data={data}></Final>
            </>
        )
    }
}

interface MinimalChildProps {
    data: any
}

export type ComponentProps = {
    iri: string
    // children: React.FC<MinimalChildProps>

    children?: React.FC<MinimalChildProps> //(props: MinimalChildProps): React.ReactNode
    component?: React.FC<MinimalChildProps>
}
export const Component: React.FC<ComponentProps> = () => {
    return <>THIS SHOULD NOT APPEAR</>
}
