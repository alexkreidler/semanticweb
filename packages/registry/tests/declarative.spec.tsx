import { Registry, Component } from "../src/index"
import { JsonLdProcessor } from "jsonld"
import React from "react"

import {
    render,
    fireEvent,
    waitFor,
    screen,
    getByText,
    act,
    waitForElement,
    findByTestId,
    getByTestId,
    waitForElementToBeRemoved,
} from "@testing-library/react"

// https://schema.org/version/latest/schemaorg-current-http.jsonld
// "@base": "http://schema.org",
// "@vocab": "http://schema.org",
const dat = {
    "@context": "http://schema.org",
    "@type": "Person",
    address: {
        "@type": "PostalAddress",
        addressLocality: "Colorado Springs",
        addressRegion: "CO",
        postalCode: "80840",
        streetAddress: "100 Main Street",
    },
    colleague: [
        "http://www.example.com/JohnColleague.html",
        "http://www.example.com/JameColleague.html",
    ],
    email: "info@example.com",
    image: "janedoe.jpg",
    jobTitle: "Research Assistant",
    name: "Jane Doe",
}

test("data expands properly", async () => {
    let exp = await JsonLdProcessor.expand(dat)
    console.log(exp)
})

// test("renders", async () => {
//     // let exp = await JsonLdProcessor.expand(dat)
//     console.log(dat)

//     const App = () => (
//         <div className="app">
//             <Registry data={dat} doExpand={true}>
//                 <Component
//                     iri="https://schema.org/Person"
//                     data-testid="person"
//                     component={({ data }) => (
//                         <div className="person">
//                             {[console.log("got data"), console.log(data)]}
//                             <h1>Person: {data.name ? data.name : "NONE"}</h1>
//                             <p>PersonP</p>
//                         </div>
//                     )}
//                 ></Component>
//                 <Component iri="https://schema.org/Article">
//                     {({ data }) => (
//                         <div>
//                             <h1>Article: {data.name}</h1>
//                         </div>
//                     )}
//                 </Component>
//             </Registry>
//         </div>
//     )
//     // We aren't doing snapshot testing RN b/c our code is highly dynamic and logic-focused

//     let { container } = render(<App></App>)
//     screen.debug()

//     let ld = getByText(container, "Loading")

//     console.log("Test completed")

//     // await waitForElementToBeRemoved(ld, {
//     //     timeout: 3000,
//     // })

//     await waitFor(
//         () => {
//             console.log("calling waiting function")
//             screen.debug()

//             expect(getByText(container, "PersonP")).toBeDefined()
//             expect(getByText(container, "Person: NONE")).toBeUndefined()
//         },
//         {
//             interval: 100,
//             timeout: 200,
//             // onT
//         }
//     )
//     // const el = await waitForElement(() => getByTestId(container, "person"))

//     // await waitFor(() => {
//     //     expect()
//     // })
//     // console.log(container)
// })
