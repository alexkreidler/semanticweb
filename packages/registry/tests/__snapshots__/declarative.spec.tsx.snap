// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  className="app"
>
  <div
    className="person"
  >
    <h1>
      Person: 
    </h1>
  </div>
</div>
`;
